docs/hugo/content/design/reconcile-interface.md

```diff
@@ -0,0 +1,227 @@
+# Reconcile Interface
+
+Today the interface is:
+
+```go
+type Reconciler interface {
+    Reconcile(
+        ctx context.Context,
+        log logr.Logger,
+        eventRecorder record.EventRecorder,
+        obj MetaObject) (ctrl.Result, error)
+}
+```
+
+This is pretty similar to the interface that controller-runtime gives us:
+
+```go
+type Reconciler interface {
+    // Reconcile performs a full reconciliation for the object referred to by the Request.
+    // The Controller will requeue the Request to be processed again if an error is non-nil or
+    // Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+    Reconcile(context.Context, Request) (Result, error)
+}
+```
+
+
+The problem is, the interface at this level means that each controller has to duplicate certain checks
+
+## What's duplicated today
+
+### Create vs Delete check
+
+```go
+    var result ctrl.Result
+    var err error
+    if !obj.GetDeletionTimestamp().IsZero() {
+        result, err = r.Delete(ctx, log, typedObj)
+    } else {
+        result, err = r.CreateOrUpdate(ctx, log, typedObj)
+    }
+
+    if err != nil {
+        return ctrl.Result{}, err
+    }
+```
+
+### Handling reconcile policy
+
+In delete: 
+```go
+    reconcilePolicy := reconcilers.GetReconcilePolicy(user, log)
+    if !reconcilePolicy.AllowsDelete() {
+        log.V(Info).Info("Bypassing delete of resource due to policy", "policy", reconcilePolicy)
+        return ctrl.Result{}, r.RemoveResourceFinalizer(ctx, log, user)
+    }
+```
+
+In create/update:
+```go
+    reconcilePolicy := reconcilers.GetReconcilePolicy(user, log)
+    if !reconcilePolicy.AllowsModify() {
+        return r.handleSkipReconcile(ctx, log, user)
+    }
+```
+
+Most of the implementation of `handleSkipReconcile` is also duplicated. One gotcha is that, in ARM resources case we want to populate status as part of that call, which requires a GET call to ARM, but in other cases we may not want to do anything.
+
+### Claiming a resource
+
+```go
+    if r.NeedToClaimResource(user) {
+        // TODO: This assumes we're requeing and doing the "stages" thing, which I am not sure if it makes sense to do for everybody anymore
+        _, err := r.ClaimResource(ctx, log, user)
+        if err != nil {
+            return ctrl.Result{}, err
+        }
+    }
+```
+
+Some claiming stuff (adding the finalizer) is universal. Other claiming stuff isn't, so we'd need a hook here.
```

Comment:
Add a bit more detail about resource-specific claiming stuff here? What extra things happen other than adding the finalizer?

Reply:
Updated wording and added some examples here.